source/slang/slang-ir-autodiff.cpp

```diff
@@ -480,29 +602,58 @@ void DifferentiableTypeConformanceContext::setFunc(IRGlobalValueWithCode* func)
                     // For differential pair types, register the differential type as well.
                     IRBuilder builder(diffPairType);
                     builder.setInsertAfter(diffPairType->getWitness());
-                    auto diffType = _lookupWitness(&builder, diffPairType->getWitness(), sharedContext->differentialAssocTypeStructKey);
-                    auto diffWitness = _lookupWitness(&builder, diffPairType->getWitness(), sharedContext->differentialAssocTypeWitnessStructKey);
-                    if (diffType && diffWitness)
-                    {
-                        differentiableWitnessDictionary.addIfNotExists((IRType*)diffType, diffWitness);
-                    }
+
+                    // TODO(sai): lot of this logic is duplicated. need to refactor.
+                    auto diffType = (diffInterfaceType == sharedContext->differentiableInterfaceType) ? 
+                        _lookupWitness(&builder, diffPairType->getWitness(), sharedContext->differentialAssocTypeStructKey, builder.getTypeKind()) :
+                        _lookupWitness(&builder, diffPairType->getWitness(), sharedContext->differentialAssocRefTypeStructKey, builder.getTypeKind());
+                    auto diffWitness = (diffInterfaceType == sharedContext->differentiableInterfaceType) ? 
+                        _lookupWitness(
+                            &builder,
+                            diffPairType->getWitness(),
+                            sharedContext->differentialAssocTypeWitnessStructKey,
+                            sharedContext->differentialAssocTypeWitnessTableType) :
+                        _lookupWitness(
+                            &builder,
+                            diffPairType->getWitness(),
+                            sharedContext->differentialAssocRefTypeWitnessStructKey,
+                            sharedContext->differentialAssocRefTypeWitnessTableType);
+                    
+                    addTypeToDictionary((IRType*)diffType, diffWitness);
                 }
             }
         }
     }
 }
 
-IRInst* DifferentiableTypeConformanceContext::lookUpConformanceForType(IRInst* type)
+IRInst* DifferentiableTypeConformanceContext::lookUpConformanceForType(IRInst* type, DiffConformanceKind kind)
 {
     IRInst* foundResult = nullptr;
-    differentiableWitnessDictionary.tryGetValue(type, foundResult);
+
```

Comment:
I still wonder why do we need two separate dictionaries. A type can either conform to valueType witness or reftype witness but not both. Why can't we consolidate the dictionaries so all lookups can be done in a single run?